### ISSUE TITLE
MethodHandle.invokeExact fails to compile on JDK 15 when targeting Java 11

### ISSUE THREAD

**Issue by srdo** [scala/bug#12236]
Here's a reproducer

https://github.com/srdo/MethodHandlesMiscompiled

The relevant code:
```
import java.lang.invoke.{ MethodHandle, MethodHandles }

object Main {
    case class TestClass(number: Int)
    def testMethod(): TestClass = {
        TestClass(10)
    }

    def wrappedMethod(handle: MethodHandle): TestClass = {
        handle.invokeExact()
    }

    def main(args: Array[String]): Unit = {
        val method = Class.forName("Main").getDeclaredMethod("testMethod")
        method.setAccessible(true)
        val handle = MethodHandles.lookup.unreflect(method)
        println(wrappedMethod(handle))
    }
}
```

Run `sbt clean compile` to see the error.

```
[error] /home/stig/MethodHandlesMiscompiled/src/main/scala/Main.scala:10:27: type mismatch;
[error]  found   : Object
[error]  required: Main.TestClass
[error]         handle.invokeExact()
[error]                           ^
[error] one error found
[error] (Compile / compileIncremental) Compilation failed
```

The code compiles on JDK 11 and also on JDK 15 if the `--release` flag is not set. I haven't been able to reproduce this in a Java project, so I suspect the bug is somewhere in the interaction between Java and Scala?

The invokeExact call is supposed to compile to bytecode like this

```
INVOKEVIRTUAL java/lang/invoke/MethodHandle.invokeExact ()LMain$TestClass;
```

The error message implies to me that the compiler fails to figure out the right return type for invokeExact when compiling it. 

Referencing the MethodHandle Javadoc https://docs.oracle.com/en/java/javase/15/docs/api/java.base/java/lang/invoke/MethodHandle.html, it has this to say about how to compile `invokeExact`:

> To issue a complete symbolic type descriptor, the compiler must also determine the return type. This is based on a cast on the method invocation expression, if there is one, or else Object if the invocation is an expression, or else void if the invocation is a statement. The cast may be to a primitive type (but not void)

For equivalent Java code 
```
var res = (TestClass)handle.invokeExact()
```
compiles to 
```
INVOKEVIRTUAL java/lang/invoke/MethodHandle.invokeExact ()LMain$TestClass;
```
which is based on the cast of the `invokeExact` expression. 

It is not possible to work around this issue with an explicit cast in Scala, since adding `asInstanceOf[TestClass]` to the expression simply moves the error to runtime, since the bytecode will still use the wrong return type for the method signature.

```
INVOKEVIRTUAL java/lang/invoke/MethodHandle.invokeExact ()Ljava/lang/Object;
CHECKCAST Main$TestClass
```

**Comment by srdo** [comment]
I should mention that I'm using Zulu 15.0.1 and Zulu 11.0.8 for this comparison. For some reason I only see the compile error when compiling using SBT directly, the code compiles in IntelliJ even when configured to delegate the build to SBT.

**Comment by srdo** [comment]
Guessing this is an SBT issue, so closing this

**Comment by SethTisue** [comment]
for the record, I suspect that there was never a problem on the sbt side and that the bug was probably fixed by either scala/scala#9530 or scala/scala#9930
